src/librustc_borrowck/borrowck/mir/elaborate_drops.rs

```diff
@@ -0,0 +1,1094 @@
+                                   MaybeUninitializedLvals::default());
+
+            match (tcx, mir, move_data) {
+                ref ctxt => ElaborateDropsCtxt {
+                    ctxt: ctxt,
+                    param_env: &param_env,
+                    flow_inits: flow_inits,
+                    flow_uninits: flow_uninits,
+                    drop_flags: FnvHashMap(),
+                    patch: MirPatch::new(mir),
+                }.elaborate()
+            }
+        };
+        pretty::dump_mir(tcx, "elaborate_drops", &0, src, mir, None);
+        elaborate_patch.apply(mir);
+        pretty::dump_mir(tcx, "elaborate_drops", &1, src, mir, None);
+    }
+}
+
+impl Pass for ElaborateDrops {}
+
+struct InitializationData {
+    live: Vec<usize>,
+    dead: Vec<usize>
+}
+
+impl InitializationData {
+    fn apply_location<'a,'tcx>(&mut self,
+                               tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                               mir: &Mir<'tcx>,
+                               move_data: &MoveData<'tcx>,
+                               loc: Location)
+    {
+        drop_flag_effects_for_location(tcx, mir, move_data, loc, |path, df| {
+            debug!("at location {:?}: setting {:?} to {:?}",
+                   loc, path, df);
+            match df {
+                DropFlagState::Live => {
+                    self.live.set_bit(path.idx());
+                    self.dead.clear_bit(path.idx());
+                }
+                DropFlagState::Dead => {
+                    self.dead.set_bit(path.idx());
+                    self.live.clear_bit(path.idx());
+                }
+            }
+        });
+    }
+
+    fn state(&self, path: MovePathIndex) -> (bool, bool) {
+        (self.live.get_bit(path.idx()), self.dead.get_bit(path.idx()))
+    }
+}
+
+impl fmt::Debug for InitializationData {
+    fn fmt(&self, _f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        Ok(())
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+/// What to do with the drop flags when we elaborate?
+enum ElaborateKind {
+    /// Clear them for this and all children
+    Normal,
+    /// Don't touch them
+    RestHead,
+    /// Clear them only for this
+    RestTail,
+}
+
+struct ElaborateDropsCtxt<'a, 'tcx: 'a> {
+    ctxt: &'a (TyCtxt<'a, 'tcx, 'tcx>, &'a Mir<'tcx>, MoveData<'tcx>),
+    param_env: &'a ty::ParameterEnvironment<'tcx>,
+    flow_inits: DataflowResults<MaybeInitializedLvals<'a, 'tcx>>,
+    flow_uninits:  DataflowResults<MaybeUninitializedLvals<'a, 'tcx>>,
+    drop_flags: FnvHashMap<MovePathIndex, u32>,
+    patch: MirPatch<'tcx>,
+}
+
+#[derive(Copy, Clone, Debug)]
+struct DropCtxt<'a, 'tcx: 'a> {
+    span: Span,
+    scope: ScopeId,
+    is_cleanup: bool,
+
+    init_data: &'a InitializationData,
+    kind: ElaborateKind,
+
+    lvalue: &'a Lvalue<'tcx>,
+    path: MovePathIndex,
+    succ: BasicBlock,
+    unwind: Option<BasicBlock>
+}
+
+impl<'b, 'tcx> ElaborateDropsCtxt<'b, 'tcx> {
```

Comment:
In general, the methods in this impl could use more comments. Ideally functions like `drop_ladder` etc would include:

1. Some relevant rust code
2. Roughly what MIR this generaets
3. What the function tries to produce